src/main/java/com/meta/chatbridge/llm/HuggingFaceLlamaPlugin.java

```diff
@@ -0,0 +1,187 @@
+//        }
+//        return tokenCount;
+        return 100;
+    }
+
+    private Optional<ArrayNode> pruneMessages(ArrayNode messages, @Nullable JsonNode functions)
+            throws JsonProcessingException {
+
+        ArrayNode output = MAPPER.createArrayNode();
+        int totalTokens = 0;
+        totalTokens += 3; // every reply is primed with <|start|>assistant<|message|>
+
+        JsonNode systemMessage = messages.get(0);
+        boolean hasSystemMessage = systemMessage.get("role").textValue().equals("system");
+        if (hasSystemMessage) {
+            // if the system message is present it's required
+            totalTokens += tokenCount(messages.get(0));
+        }
+        for (int i = messages.size() - 1; i >= 0; i--) {
+            JsonNode m = messages.get(i);
+            String role = m.get("role").textValue();
+            if (role.equals("system")) {
+                continue; // system has already been counted
+            }
+            totalTokens += tokenCount(m);
+            if (totalTokens > config.maxInputTokens()) {
+                break;
+            }
+            output.insert(0, m);
+        }
+        if (hasSystemMessage) {
+            output.insert(0, systemMessage);
+        }
+
+        if ((hasSystemMessage && output.size() <= 1) || output.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(output);
+    }
+
+    @Override
+    public T handle(MessageStack<T> messageStack) throws IOException {
+        T fromUser = messageStack.tail();
+
+        ObjectNode body = MAPPER.createObjectNode();
+        ObjectNode params = MAPPER.createObjectNode();
+
+        config.topP().ifPresent(v -> params.put("top_p", v));
+        config.temperature().ifPresent(v -> params.put("temperature", v));
+        config.maxOutputTokens().ifPresent(v -> params.put("max_new_tokens", v));
+
+        body.set("parameters", params);
+
+        String prompt = getPrompt(messageStack);
+
+        body.put("inputs", prompt);
+
+        String bodyString;
+        try {
+            bodyString = MAPPER.writeValueAsString(body);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e); // this should be impossible
+        }
+        Response response =
+                Request.post(endpoint)
+                        .bodyString(bodyString, ContentType.APPLICATION_JSON)
+                        .setHeader("Authorization", "Bearer " + config.apiKey())
+                        .execute();
+
+        JsonNode responseBody = MAPPER.readTree(response.returnContent().asBytes());
+        String allGeneratedText = responseBody.get(0).get("generated_text").textValue();
+        String llmResponse = allGeneratedText.trim().replace(prompt.trim(), "");
+        Instant timestamp = Instant.now();
+
+        return messageStack.newMessageFromBot(timestamp, llmResponse);
+    }
+
+    private String getPrompt(MessageStack<T> MessageStack) {
```

Comment:
`get`/`set` prefixes are a convention for methods returning a value from the internals of an object (getter/setter). While this probably isn't technically wrong it's a weird use of `get`. `createPrompt` or something similar would probably be better.

also I've been using a convention that drops the `get`/`set` prefixes for getters and setters as you can infer if a method is a getter or setter from it's signature and I think it looks cleaner :)

Reply:
This prompt creation logic would belong better in a `PromptBuilder` class that could also handle the pruning. Don't worry about this right now though. (honestly this is probably true for the OpenAI plugin as well)